File: app/src/main/java/yhb/chorus/main/MainFragment.java
package yhb.chorus.main;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import androidx.annotation.Nullable;
import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import androidx.fragment.app.Fragment;
import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import yhb.chorus.R;
import yhb.chorus.common.adapter.SimpleAdapter;
import yhb.chorus.common.adapter.base.SimpleHolder;
import yhb.chorus.entity.MP3;
import yhb.chorus.list.ListActivity;
import yhb.chorus.utils.ActivityUtils;

import static yhb.chorus.main.MainActivity.TAG;
import static yhb.chorus.service.PlayCenter.MODE_LIST_LOOP;
import static yhb.chorus.service.PlayCenter.MODE_RANDOM;
import static yhb.chorus.service.PlayCenter.MODE_SINGLE_LOOP;

/**
 * Created by yhb on 18-1-17.
 */

public class MainFragment extends Fragment implements MainContract.View, View.OnClickListener {

    private MainContract.Presenter mPresenter;
    private ViewPager mCoverViewPager;
    private ImageButton buttonPlayOrPause;
    private ImageButton buttonPlayMode;
    private TextView textViewCurrentProgress, textViewMaxProgress, textViewSongName, textViewArtistName;
    private SeekBar mSeekBar, mSeekBarVolume, mSeekBarVolumeSystem;
    private PageSelectedListener mPageSelectedListener;
    private Runnable invalidateConsole;
    private Handler handler;
    private int mCoverSize = -1;

    public static MainFragment newInstance() {
        Bundle args = new Bundle();
        MainFragment fragment = new MainFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        super.onCreate(savedInstanceState);

        handler = new Handler();
        invalidateConsole = new Runnable() {
            @Override
            public void run() {
                try {
                    mPresenter.reloadConsoleData();
                    handler.postDelayed(invalidateConsole, 500);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        };

        mCoverSize = ActivityUtils.getScreenWidth(getActivity()) * 4 / 5;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_main, container, false);
        bindViews(root);
        mPresenter.start();
        return root;
    }

    @Override
    public void onResume() {
        super.onResume();
        mPresenter.reloadCurrentWidgetsData();
        mPresenter.loadSavedSetting();
        mPresenter.loadCoversAsync();
        invalidateSeekBarVolumeSystem(
                mPresenter.getCurrentVolumeSystem(),
                mPresenter.getMaxVolumeSystem()
        );

        handler.postDelayed(invalidateConsole, 500);

    }

    private void bindViews(View root) {
        ImageButton buttonNext = root.findViewById(R.id.image_button_next);
        ImageButton buttonPrevious = root.findViewById(R.id.image_button_previous);
        buttonPlayMode = root.findViewById(R.id.image_button_play_mode);
        ImageButton buttonQueueMusic = root.findViewById(R.id.image_button_queue_music);
        buttonPlayOrPause = root.findViewById(R.id.image_button_play_or_pause);
        mCoverViewPager = root.findViewById(R.id.view_pager_cover);
        mSeekBar = root.findViewById(R.id.slim_seek_bar);
        mSeekBarVolume = root.findViewById(R.id.slim_seek_bar_volume);
        mSeekBarVolumeSystem = root.findViewById(R.id.slim_seek_bar_volume_system);
        textViewMaxProgress = root.findViewById(R.id.text_view_max_progress);
        textViewCurrentProgress = root.findViewById(R.id.text_view_current_progress);
        textViewSongName = root.findViewById(R.id.text_view_song_name);
        textViewArtistName = root.findViewById(R.id.text_view_artist_name);
        mSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            private int progress = -1;

            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {
                    this.progress = progress;
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                if (progress != -1) {
                    mPresenter.seekTo(progress);
                    progress = -1;
                }
            }
        });

        mSeekBarVolume.setMax(10);
        mSeekBarVolume.setProgress(0);
        mSeekBarVolume.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {

            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {
                    float volume = ((float) progress) / ((float) seekBar.getMax());
                    mPresenter.setVolume(volume);
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });

        mSeekBarVolumeSystem.setMax(mPresenter.getMaxVolumeSystem());
        mSeekBarVolumeSystem.setProgress(mPresenter.getCurrentVolumeSystem());
        mSeekBarVolumeSystem.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {

            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {
                    mPresenter.setVolumeSystem(progress);
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });



        ViewGroup.LayoutParams layoutParams = mCoverViewPager.getLayoutParams();
        int coverSize = getCoverSize();
        layoutParams.width = coverSize;
        layoutParams.height = coverSize;

        CoverPagerAdapter coverPagerAdapter = new CoverPagerAdapter();
        int initialIndex = coverPagerAdapter.getCount() / 4 * 3 + 1;
        mPageSelectedListener = new PageSelectedListener(
                initialIndex, new OnSelectedListener() {
            @Override
            public void onNext() {
                mPresenter.next();
                mPresenter.reloadCurrentWidgetsData();
                mPresenter.loadCoversAsync();
                Log.d(TAG, "onNext: ");
            }

            @Override
            public void onPrevious() {
                Log.d(TAG, "onPrevious: ");
                mPresenter.previous();
                mPresenter.reloadCurrentWidgetsData();
                mPresenter.loadCoversAsync();
            }
        }
        );
        mCoverViewPager.setLayoutParams(layoutParams);
        mCoverViewPager.setAdapter(coverPagerAdapter);
        mCoverViewPager.setCurrentItem(initialIndex);
        mCoverViewPager.addOnPageChangeListener(mPageSelectedListener);


        buttonNext.setOnClickListener(this);
        buttonPrevious.setOnClickListener(this);
        buttonPlayOrPause.setOnClickListener(this);
        buttonPlayMode.setOnClickListener(this);
        buttonQueueMusic.setOnClickListener(this);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mPresenter.release();
        handler.removeCallbacks(invalidateConsole);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.main_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.sort_out:
                Intent intent = ListActivity.newIntent(getActivity());
                startActivity(intent);
                break;
            default:
                break;
        }
        return true;
    }

    @Override
    public void setPresenter(MainContract.Presenter presenter) {
        this.mPresenter = presenter;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.image_button_next:
                mPresenter.next();
                mPresenter.reloadCurrentWidgetsData();
                mPresenter.loadCoversAsync();
                break;
            case R.id.image_button_play_or_pause:
                mPresenter.playOrPause();
                mPresenter.reloadCurrentWidgetsData();
                mPresenter.loadCoversAsync();
                break;
            case R.id.image_button_previous:
                mPresenter.previous();
                mPresenter.reloadCurrentWidgetsData();
                mPresenter.loadCoversAsync();
                break;
            case R.id.image_button_play_mode:
                mPresenter.nextPlayMode();
                mPresenter.reloadCurrentWidgetsData();
                break;
            case R.id.image_button_queue_music:
                showBottomSheet();
                break;
        }
    }

    @Override
    public void invalidateSeekBarVolumeSystem(int currentVolume, int volumeSystemMax) {
        mSeekBarVolumeSystem.setMax(volumeSystemMax);
        mSeekBarVolumeSystem.setProgress(currentVolume);
    }

    @Override
    public void invalidateCovers(final Bitmap[] bitmaps) {
        getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {

                if (bitmaps != null) {
                    ImageView[] covers = ((CoverPagerAdapter) mCoverViewPager.getAdapter()).getCovers();

                    int i = mPageSelectedListener.currentIndex();
                    covers[i].setImageBitmap(bitmaps[1]);

                    switch (i) {
                        case 0:
                            covers[2].setImageBitmap(bitmaps[0]);
                            covers[1].setImageBitmap(bitmaps[2]);
                            break;
                        case 1:
                            covers[0].setImageBitmap(bitmaps[0]);
                            covers[2].setImageBitmap(bitmaps[2]);
                            break;
                        case 2:
                            covers[1].setImageBitmap(bitmaps[0]);
                            covers[0].setImageBitmap(bitmaps[2]);
                            break;
                    }

                }
            }
        });
    }

    @Override
    public void invalidateWidgets(int progress, int playMode, String songName, String artistName) {

        mSeekBarVolume.setProgress(progress);

        switch (playMode) {
            case MODE_LIST_LOOP:
                buttonPlayMode.setImageResource(R.drawable.ic_repeat_list);
                break;
            case MODE_RANDOM:
                buttonPlayMode.setImageResource(R.drawable.ic_shuffle);
                break;
            case MODE_SINGLE_LOOP:
                buttonPlayMode.setImageResource(R.drawable.ic_repeat_one);
                break;
            default:
                buttonPlayMode.setImageResource(R.drawable.ic_repeat_list);
                break;
        }

        textViewSongName.setText(songName);

        textViewArtistName.setText(artistName);
    }

    @Override
    public void invalidatePlayStatus(boolean playing, int progress) {
        MP3 currentMP3 = mPresenter.getCurrentMP3();
        if (currentMP3 == null) {
            return;
        }
        int maxProgress = currentMP3.getDuration();

        mSeekBar.setMax(maxProgress);
        mSeekBar.setProgress(progress);

        textViewMaxProgress.setText(mm2min(maxProgress));
        textViewCurrentProgress.setText(mm2min(progress));

        if (playing) {
            buttonPlayOrPause.setImageResource(R.drawable.ic_pause_circle_outline);
        } else {
            buttonPlayOrPause.setImageResource(R.drawable.ic_play_circle_outline);
        }

    }

    @Override
    public int getCoverSize() {
        return mCoverSize;
    }

    private BottomSheetDialog bottomSheetDialog = null;

    private SimpleAdapter<MP3> mQueueMP3SimpleAdapter;

    private void showBottomSheet() {
        if (bottomSheetDialog != null && mQueueMP3SimpleAdapter != null) {
            mQueueMP3SimpleAdapter.performDataSetChanged(mPresenter.loadQueueMP3sFromMemory());
            bottomSheetDialog.show();
            return;
        }
        mQueueMP3SimpleAdapter = new SimpleAdapter<MP3>(getActivity(), R.layout.item_mp3_simple) {
            @Override
            public void convert(SimpleHolder holder, final MP3 mp3) {
                TextView textView = holder.getView(R.id.text_view_song_name);
                textView.setText(String.format("%s / %s", mp3.getTitle(), mp3.getArtist()));
                textView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        mPresenter.point(mp3);
                        mPresenter.reloadCurrentWidgetsData();
                        mPresenter.loadCoversAsync();
                    }
                });
            }
        };
        RecyclerView recyclerView = (RecyclerView) LayoutInflater.from(getActivity())
                .inflate(R.layout.content_queue_song, null);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerView.setAdapter(mQueueMP3SimpleAdapter);
        mQueueMP3SimpleAdapter.performDataSetChanged(mPresenter.loadQueueMP3sFromMemory());

        bottomSheetDialog = new BottomSheetDialog(getActivity());
        bottomSheetDialog.setCancelable(true);
        bottomSheetDialog.setCanceledOnTouchOutside(true);
        bottomSheetDialog.setContentView(recyclerView);

        // 曲线救国设置 peekHeight（出现时的高度）
        int maxHeight = ActivityUtils.getScreenHeight(getActivity()) * 3 / 5;
        View view = bottomSheetDialog.getWindow().findViewById(com.google.android.material.R.id.design_bottom_sheet);
        BottomSheetBehavior.from(view).setPeekHeight(maxHeight);

        bottomSheetDialog.show();

        // 设置最大高度
        bottomSheetDialog.getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, maxHeight);
        // 解决设置最大高度后的悬空问题
        bottomSheetDialog.getWindow().setGravity(Gravity.BOTTOM);

    }

    private String mm2min(int mm) {
        String min, sec;
        if (mm / 1000 / 60 > 9) {
            min = mm / 1000 / 60 + "";
        } else {
            min = "0" + mm / 1000 / 60;
        }

        if (((mm / 1000) % 60) <= 9) {
            sec = "0" + (mm / 1000) % 60;
        } else {
            sec = (mm / 1000) % 60 + "";
        }
        return min + ":" + sec;
    }

    class CoverPagerAdapter extends PagerAdapter {


        private ImageView covers[];

        ImageView[] getCovers() {
            return covers;
        }

        CoverPagerAdapter() {
            covers = new ImageView[3];

            ImageView cover0, cover1, cover2;

            cover0 = new ImageView(MainFragment.this.getActivity());
            cover1 = new ImageView(MainFragment.this.getActivity());
            cover2 = new ImageView(MainFragment.this.getActivity());

            covers[0] = cover0;
            covers[1] = cover1;
            covers[2] = cover2;

            int screenWidth = ActivityUtils.getScreenWidth(getActivity());
            for (ImageView cover : covers) {
                cover.setBackgroundColor(Color.parseColor("#44888888"));
                cover.setPadding(screenWidth / 50, screenWidth / 50, screenWidth / 50, screenWidth / 50);
                cover.setScaleType(ImageView.ScaleType.FIT_XY);
            }


        }

        @Override
        public Object instantiateItem(ViewGroup container, int position) {
            View view = covers[position % 3];
            if (container.equals(view.getParent())) {
                container.removeView(view);
            }
            container.addView(view);
            return view;
        }

        @Override
        public void destroyItem(ViewGroup container, int position, Object object) {

        }

        @Override
        public int getCount() {
            return Integer.MAX_VALUE / 2;
        }

        @Override
        public int getItemPosition(Object object) {
            return POSITION_NONE;
        }

        @Override
        public boolean isViewFromObject(View view, Object object) {
            return view == object;
        }

    }

    interface OnSelectedListener {

        void onNext();

        void onPrevious();

    }

    class PageSelectedListener implements ViewPager.OnPageChangeListener {

        private int oldIndex;

        private OnSelectedListener mOnSelectedListener;

        PageSelectedListener(int initialIndex, OnSelectedListener onSelectedListener) {
            this.oldIndex = initialIndex;
            mOnSelectedListener = onSelectedListener;
        }

        @Override
        public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

        }

        @Override
        public void onPageSelected(int position) {
            if (position == oldIndex + 1) {
                mOnSelectedListener.onNext();
            } else if (position == oldIndex - 1) {
                mOnSelectedListener.onPrevious();
            }

            oldIndex = position;
        }

        @Override
        public void onPageScrollStateChanged(int state) {

        }

        int currentIndex() {
            return oldIndex % 3;
        }

    }

}

File: app/src/main/java/yhb/chorus/main/MainFragment.kt
package yhb.chorus.main

import android.graphics.Bitmap
import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.*
import android.widget.ImageView
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import yhb.chorus.R
import yhb.chorus.common.adapter.SimpleAdapter
import yhb.chorus.common.adapter.base.SimpleHolder
import yhb.chorus.databinding.FragmentMainBinding
import yhb.chorus.entity.MP3
import yhb.chorus.list.ListActivity
import yhb.chorus.service.PlayCenter.MODE_LIST_LOOP
import yhb.chorus.service.PlayCenter.MODE_RANDOM
import yhb.chorus.service.PlayCenter.MODE_SINGLE_LOOP
import yhb.chorus.utils.ActivityUtils

/**
 * Created by yhb on 18-1-17.
 */
class MainFragment : Fragment(), MainContract.View, View.OnClickListener {

    private lateinit var mainPresenter: MainContract.Presenter
    private lateinit var invalidateConsole: Runnable
    private lateinit var pageSelectedListener: PageSelectedListener

    private var _binding: FragmentMainBinding? = null
    private val binding: FragmentMainBinding get() = _binding!!

    private val handler = Handler()
    private var coverSize = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        setHasOptionsMenu(true)
        super.onCreate(savedInstanceState)
        invalidateConsole = Runnable {
            try {
                mainPresenter.reloadConsoleData()
                handler.postDelayed(invalidateConsole, 500)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        coverSize = ActivityUtils.getScreenWidth(activity) * 4 / 5
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentMainBinding.inflate(inflater, container, false)
        bindViews(binding)
        mainPresenter.start()
        return binding.root
    }

    override fun onResume() {
        super.onResume()
        mainPresenter.reloadCurrentWidgetsData()
        mainPresenter.loadSavedSetting()
        mainPresenter.loadCoversAsync()
        invalidateSeekBarVolumeSystem(
                mainPresenter.currentVolumeSystem,
                mainPresenter.maxVolumeSystem
        )
        handler.postDelayed(invalidateConsole, 500)
    }

    private fun bindViews(binding: FragmentMainBinding) {
        binding.slimSeekBar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            private var progress = -1
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    this.progress = progress
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {
                if (progress != -1) {
                    mainPresenter.seekTo(progress)
                    progress = -1
                }
            }
        })
        binding.slimSeekBarVolume.max = 10
        binding.slimSeekBarVolume.progress = 0
        binding.slimSeekBarVolume.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    val volume = progress.toFloat() / seekBar.max.toFloat()
                    mainPresenter.setVolume(volume)
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
        binding.slimSeekBarVolumeSystem.max = mainPresenter.maxVolumeSystem
        binding.slimSeekBarVolumeSystem.progress = mainPresenter.currentVolumeSystem
        binding.slimSeekBarVolumeSystem.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    mainPresenter.setVolumeSystem(progress)
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
        val layoutParams = binding.viewPagerCover.layoutParams
        val coverSize = coverSize
        layoutParams.width = coverSize
        layoutParams.height = coverSize
        val coverPagerAdapter = CoverPagerAdapter()
        val initialIndex = coverPagerAdapter.count / 4 * 3 + 1
        pageSelectedListener = PageSelectedListener(
                initialIndex, object : OnSelectedListener {
            override fun onNext() {
                mainPresenter.next()
                mainPresenter.reloadCurrentWidgetsData()
                mainPresenter.loadCoversAsync()
                Log.d(MainActivity.TAG, "onNext: ")
            }

            override fun onPrevious() {
                Log.d(MainActivity.TAG, "onPrevious: ")
                mainPresenter.previous()
                mainPresenter.reloadCurrentWidgetsData()
                mainPresenter.loadCoversAsync()
            }
        }
        )
        binding.viewPagerCover.layoutParams = layoutParams
        binding.viewPagerCover.adapter = coverPagerAdapter
        binding.viewPagerCover.currentItem = initialIndex
        binding.viewPagerCover.addOnPageChangeListener(pageSelectedListener)
        binding.imageButtonNext.setOnClickListener(this)
        binding.imageButtonPrevious.setOnClickListener(this)
        binding.imageButtonPlayOrPause.setOnClickListener(this)
        binding.imageButtonPlayMode.setOnClickListener(this)
        binding.imageButtonQueueMusic.setOnClickListener(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        mainPresenter.release()
        handler.removeCallbacks(invalidateConsole)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.main_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.sort_out -> {
                val intent = ListActivity.newIntent(activity)
                startActivity(intent)
            }
            else -> {
            }
        }
        return true
    }

    override fun setPresenter(presenter: MainContract.Presenter) {
        mainPresenter = presenter
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.image_button_next -> {
                mainPresenter.next()
                mainPresenter.reloadCurrentWidgetsData()
                mainPresenter.loadCoversAsync()
            }
            R.id.image_button_play_or_pause -> {
                mainPresenter.playOrPause()
                mainPresenter.reloadCurrentWidgetsData()
                mainPresenter.loadCoversAsync()
            }
            R.id.image_button_previous -> {
                mainPresenter.previous()
                mainPresenter.reloadCurrentWidgetsData()
                mainPresenter.loadCoversAsync()
            }
            R.id.image_button_play_mode -> {
                mainPresenter.nextPlayMode()
                mainPresenter.reloadCurrentWidgetsData()
            }
            R.id.image_button_queue_music -> showBottomSheet()
        }
    }

    override fun invalidateSeekBarVolumeSystem(currentVolume: Int, volumeSystemMax: Int) {
        binding.slimSeekBarVolumeSystem.max = volumeSystemMax
        binding.slimSeekBarVolumeSystem.progress = currentVolume
    }

    override fun invalidateCovers(bitmaps: Array<Bitmap>) {
        activity?.runOnUiThread {
            val covers = (binding.viewPagerCover.adapter as CoverPagerAdapter).covers
            val index = pageSelectedListener.currentIndex()
            covers[index].setImageBitmap(bitmaps[1])
            when (index) {
                0 -> {
                    covers[2].setImageBitmap(bitmaps[0])
                    covers[1].setImageBitmap(bitmaps[2])
                }
                1 -> {
                    covers[0].setImageBitmap(bitmaps[0])
                    covers[2].setImageBitmap(bitmaps[2])
                }
                2 -> {
                    covers[1].setImageBitmap(bitmaps[0])
                    covers[0].setImageBitmap(bitmaps[2])
                }
            }
        }
    }

    override fun invalidateWidgets(progress: Int, playMode: Int, songName: String, artistName: String) {
        binding.slimSeekBarVolume.progress = progress
        when (playMode) {
            MODE_LIST_LOOP -> binding.imageButtonPlayMode.setImageResource(R.drawable.ic_repeat_list)
            MODE_RANDOM -> binding.imageButtonPlayMode.setImageResource(R.drawable.ic_shuffle)
            MODE_SINGLE_LOOP -> binding.imageButtonPlayMode.setImageResource(R.drawable.ic_repeat_one)
            else -> binding.imageButtonPlayMode.setImageResource(R.drawable.ic_repeat_list)
        }
        binding.textViewSongName.text = songName
        binding.textViewArtistName.text = artistName
    }

    override fun invalidatePlayStatus(playing: Boolean, progress: Int) {
        val currentMP3 = mainPresenter.currentMP3 ?: return
        val maxProgress = currentMP3.duration
        binding.slimSeekBar.max = maxProgress
        binding.slimSeekBar.progress = progress
        binding.textViewMaxProgress.text = mm2min(maxProgress)
        binding.textViewCurrentProgress.text = mm2min(progress)
        if (playing) {
            binding.imageButtonPlayOrPause.setImageResource(R.drawable.ic_pause_circle_outline)
        } else {
            binding.imageButtonPlayOrPause.setImageResource(R.drawable.ic_play_circle_outline)
        }
    }

    override fun getCoverSize(): Int {
        return coverSize
    }

    private var bottomSheetDialog: BottomSheetDialog? = null
    private val queueMP3SimpleAdapter: SimpleAdapter<MP3> by lazy {
        object : SimpleAdapter<MP3>(requireNotNull(activity), R.layout.item_mp3_simple) {
            override fun convert(holder: SimpleHolder, mp3: MP3) {
                val textView = holder.getView<TextView>(R.id.text_view_song_name)
                textView.text = String.format("%s / %s", mp3.title, mp3.artist)
                textView.setOnClickListener {
                    mainPresenter.point(mp3)
                    mainPresenter.reloadCurrentWidgetsData()
                    mainPresenter.loadCoversAsync()
                }
            }
        }
    }

    private fun showBottomSheet() {
        var sheetDialog = bottomSheetDialog
        if (sheetDialog != null) {
            queueMP3SimpleAdapter.performDataSetChanged(mainPresenter.loadQueueMP3sFromMemory())
            sheetDialog.show()
            return
        }
        val activity = activity ?: return
        val recyclerView = LayoutInflater.from(activity)
                .inflate(R.layout.content_queue_song, binding.root, false) as RecyclerView
        recyclerView.layoutManager = LinearLayoutManager(activity)
        recyclerView.adapter = queueMP3SimpleAdapter

        queueMP3SimpleAdapter.performDataSetChanged(mainPresenter.loadQueueMP3sFromMemory())

        sheetDialog = BottomSheetDialog(activity).apply {
            setCancelable(true)
            setCanceledOnTouchOutside(true)
            setContentView(recyclerView)
            bottomSheetDialog = sheetDialog
        }

        val window = sheetDialog.window ?: return
        // 曲线救国设置 peekHeight（出现时的高度）
        val maxHeight = ActivityUtils.getScreenHeight(activity) * 3 / 5
        val view = window.findViewById<View>(com.google.android.material.R.id.design_bottom_sheet)
        BottomSheetBehavior.from(view).peekHeight = maxHeight
        sheetDialog.show()

        // 设置最大高度
        window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, maxHeight)
        // 解决设置最大高度后的悬空问题
        window.setGravity(Gravity.BOTTOM)

    }

    private fun mm2min(mm: Int): String {
        val min: String = if (mm / 1000 / 60 > 9) {
            (mm / 1000 / 60).toString()
        } else {
            "0" + mm / 1000 / 60
        }
        val sec: String = if (mm / 1000 % 60 <= 9) {
            "0" + mm / 1000 % 60
        } else {
            ((mm / 1000) % 60).toString()
        }
        return "$min:$sec"
    }

    internal inner class CoverPagerAdapter : PagerAdapter() {

        val covers: Array<ImageView> = Array(3) { ImageView(this@MainFragment.activity) }

        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val view: View = covers[position % 3]
            if (container == view.parent) {
                container.removeView(view)
            }
            container.addView(view)
            return view
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {}
        override fun getCount(): Int {
            return Int.MAX_VALUE / 2
        }

        override fun getItemPosition(`object`: Any): Int {
            return POSITION_NONE
        }

        override fun isViewFromObject(view: View, `object`: Any): Boolean {
            return view === `object`
        }

        init {
            val screenWidth = ActivityUtils.getScreenWidth(activity)
            for (cover in covers) {
                cover.setBackgroundColor(Color.parseColor("#44888888"))
                cover.setPadding(screenWidth / 50, screenWidth / 50, screenWidth / 50, screenWidth / 50)
                cover.scaleType = ImageView.ScaleType.FIT_XY
            }
        }
    }

    internal interface OnSelectedListener {
        fun onNext()
        fun onPrevious()
    }

    internal inner class PageSelectedListener(private var oldIndex: Int, private val mOnSelectedListener: OnSelectedListener) : OnPageChangeListener {
        override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
        override fun onPageSelected(position: Int) {
            if (position == oldIndex + 1) {
                mOnSelectedListener.onNext()
            } else if (position == oldIndex - 1) {
                mOnSelectedListener.onPrevious()
            }
            oldIndex = position
        }

        override fun onPageScrollStateChanged(state: Int) {}
        fun currentIndex(): Int {
            return oldIndex % 3
        }

    }

    companion object {

        @JvmStatic
        fun newInstance(): MainFragment {
            val args = Bundle()
            val fragment = MainFragment()
            fragment.arguments = args
            return fragment
        }
    }
}